llvm/lib/Transforms/Scalar/IndVarSimplify.cpp

```diff
@@ -806,12 +807,9 @@ static bool isLoopCounter(PHINode* Phi, Loop *L,
   if (!SE->isSCEVable(Phi->getType()))
     return false;
 
-  const SCEVAddRecExpr *AR = dyn_cast<SCEVAddRecExpr>(SE->getSCEV(Phi));
-  if (!AR || AR->getLoop() != L || !AR->isAffine())
-    return false;
-
-  const SCEV *Step = dyn_cast<SCEVConstant>(AR->getStepRecurrence(*SE));
-  if (!Step || !Step->isOne())
+  const SCEV *S = SE->getSCEV(Phi);
+  if (!match(S, m_scev_AffineAddRec(m_SCEV(), m_scev_One())) ||
+      cast<SCEVAddRecExpr>(S)->getLoop() != L)
```

Comment:
Might make sense to have a m_scev_AffineAddRec variant that takes the Loop as an argument?

Reply:
Yeah I had some code for that lying around somewhere, let me see if I can find it again

Reply:
Will do in a follow-up, if @fhahn doesn't beat me to it.